azurerm/internal/services/logic/integration_service_environment.go

```diff
@@ -0,0 +1,450 @@
+	for _, item := range input {
+		results = append(results, logic.ResourceReference{
+			ID: utils.String(item.(string)),
+		})
+	}
+	return &results
+}
+
+func flattenSubnetResourceID(input *[]logic.ResourceReference) []interface{} {
+	subnetIDs := make([]interface{}, 0)
+	if input == nil {
+		return subnetIDs
+	}
+
+	for _, resourceRef := range *input {
+		if resourceRef.ID == nil || *resourceRef.ID == "" {
+			continue
+		}
+
+		subnetIDs = append(subnetIDs, resourceRef.ID)
+	}
+
+	return subnetIDs
+}
+
+func getSubnetIDs(input *logic.IntegrationServiceEnvironment) []interface{} {
+	emptySubnetIDs := make([]interface{}, 0)
+	if input == nil {
+		return emptySubnetIDs
+	}
+
+	if props := input.Properties; props != nil {
+		if netCfg := props.NetworkConfiguration; netCfg != nil {
+			return flattenSubnetResourceID(netCfg.Subnets)
+		}
+	}
+
+	return emptySubnetIDs
+}
+
+func integrationServiceEnvironmentDeleteStateRefreshFunc(ctx context.Context, client *clients.Client, iseID string, subnetIDs []interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		linkExists, err := linkExists(ctx, client, iseID, subnetIDs)
+		if err != nil {
+			return string(logic.WorkflowProvisioningStateDeleting), string(logic.WorkflowProvisioningStateDeleting), err
+		}
+
+		if linkExists {
+			return string(logic.WorkflowProvisioningStateDeleting), string(logic.WorkflowProvisioningStateDeleting), nil
+		}
+
+		return string(logic.WorkflowProvisioningStateDeleted), string(logic.WorkflowProvisioningStateDeleted), nil
+	}
+}
+
+func linkExists(ctx context.Context, client *clients.Client, iseID string, subnetIDs []interface{}) (bool, error) {
+	for _, subnetID := range subnetIDs {
+		id := *(subnetID.(*string))
+		log.Printf("Checking links on subnetID: %q\n", id)
+
+		hasLink, err := serviceAssociationLinkExists(ctx, client.Network.ServiceAssociationLinksClient, iseID, id)
+		if err != nil {
+			return false, err
+		}
+
+		if hasLink {
+			return true, nil
+		} else {
+			hasLink, err := resourceNavigationLinkExists(ctx, client.Network.ResourceNavigationLinksClient, id)
+			if err != nil {
+				return false, err
+			}
+
+			if hasLink {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func serviceAssociationLinkExists(ctx context.Context, client *network.ServiceAssociationLinksClient, iseID string, subnetID string) (bool, error) {
+	id, err := networkParse.SubnetID(subnetID)
+	if err != nil {
+		return false, err
+	}
+
+	resp, err := client.List(ctx, id.ResourceGroup, id.VirtualNetworkName, id.Name)
+
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return false, nil
+		}
+		return false, fmt.Errorf("listing Service Association Links from Virtual Network %q, subnet %q (Resource Group %q): %+v", id.VirtualNetworkName, id.Name, id.ResourceGroup, err)
+	}
+
+	if resp.Value != nil {
+		for _, link := range *resp.Value {
+			if strings.EqualFold(iseID, *link.ServiceAssociationLinkPropertiesFormat.Link) {
```

Comment:
there's a crash here if `link.ServiceAssociationLinkPropertiesFormat` or `link.ServiceAssociationLinkPropertiesFormat.Link` is nil - could we nil-check this?

Reply:
Thanks, good catch. Will change it.